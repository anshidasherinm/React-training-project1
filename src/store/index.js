import { configureStore } from "@reduxjs/toolkit";
import todoSlice from "store/todo/todo-slice";
import labelSlice from "store/label/label-slice";

const store = configureStore({
  reducer: {
    todo: todoSlice.reducer,
    label: labelSlice.reducer,
  },
});

export default store;
