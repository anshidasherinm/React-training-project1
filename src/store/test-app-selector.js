const state = {
  todo: {
    items: [
      {
        checked: false,
        id: 714,
        key: "-NOqX_ZRWmYBbJS0hkJh",
        labelKey: "-NOqXGK_kreZB5AW_PFG",
        title: "study for exam",
      },
    ],
    allItems: [],
    addingTodoItem: false,
    addingTodoItemError: false,
    loadingList: false,
    loadingListError: false,
    updatingTodo: false,
    updatingTodoError: false,
    deletingTodoItem: false,
    deletingTodoItemError: false,
  },
  label: {
    labels: [{ id: 1463, key: "-NOqXGK_kreZB5AW_PFG", name: "School" }],
    addingLabel: false,
    addingLabelError: false,
    loadingLabels: false,
    loadingLabelError: false,
    updatingLabel: false,
    updatingLabelError: false,
    deletingLabel: false,
    deletingLabelError: false,
  },
};

export const testUseAppSelector = (f) => f(state);
