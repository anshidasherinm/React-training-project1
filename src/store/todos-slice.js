import { createSlice } from "@reduxjs/toolkit";

const todoSlice = createSlice({
  name: "todos",
  initialState: {
    items: [],
    changed: false,
  },
  reducers: {
    replaceTodos(state, action) {
      state.items = action.payload.items;
    },

    addItemtodo(state, action) {
      const newItem = action.payload;
      state.changed = true;
      state.items = [...state.items, newItem];
      // console.log(state.items);//
    },
    removeItemFromTodos(state, action) {
      const id = action.payload;
      const items = state.items.filter((item) => item.id !== id);
      state.items = items;
      state.changed = true;
    },
    addCheckHandler(state, action) {
      const id = action.payload;
      // console.log(id);
      const newTodoList = state.items.map((todo) => {
        if (todo.id === id) {
          return { ...todo, checked: !todo.checked };
        }
        return todo;
      });
      state.items = newTodoList;
      state.changed = true;
    },
  },
});

export const todoActions = todoSlice.actions;
export default todoSlice;
