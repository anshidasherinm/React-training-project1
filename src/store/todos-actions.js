import { uiActions } from "./ui-slice";
import { todoActions } from "./todos-slice";
export const fetchTodoData = () => {
  return async (dispatch) => {
    const fetchData = async () => {
      const response = await fetch(
        "https://react-training-ad656-default-rtdb.firebaseio.com/todos.json"
      );

      if (!response.ok) {
        throw new Error("Could not fetch cart data!");
      }

      const data = await response.json();
      // console.log(data.items);
      return data;
    };

    try {
      const toDoDatas = await fetchData();
      // console.log(toDoDatas);
      if (toDoDatas === null) {
        throw new Error("Empty todo List");
      }
      dispatch(
        todoActions.replaceTodos({
          items: toDoDatas.items || [],
        })
      );
    } catch (error) {
      dispatch(
        uiActions.showNotification({
          status: error.message === "Empty todo List" ? "success" : "error",
          title: error.message === "Empty todo List" ? "Empty" : "Error!",
          message: error.message,
        })
      );
    }
  };
};

export const sendTodoData = (todo) => {
  //action craetor thunk
  return async (dispatch) => {
    dispatch(
      uiActions.showNotification({
        status: "pending",
        title: "Updating...",
        message: "Updating todo data! ",
      })
    );

    const sendRequest = async () => {
      const response = await fetch(
        "https://react-training-ad656-default-rtdb.firebaseio.com/todos.json",
        {
          method: "PUT",
          body: JSON.stringify({
            items: todo.items,
            // changed: cart.changed,
          }),
        }
      );
      if (!response.ok) {
        throw new Error("Updating todo data failed.");
      }
    };
    try {
      await sendRequest();
      dispatch(
        uiActions.showNotification({
          status: "success",
          title: "Success !",
          message: "Updated todo data successfully ",
        })
      );
    } catch (error) {
      dispatch(
        uiActions.showNotification({
          status: "error",
          title: "Error!",
          message: error.message,
        })
      );
    }
  };
};
