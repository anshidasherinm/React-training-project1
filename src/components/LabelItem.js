import React, { useMemo, useRef, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { deleteLabel, updateLabel } from "../store/label/label-thunk";
import { todoActions } from "../store/todo/todo-slice";
import { Button } from "@mui/material";
import IconMenu from "./IconMenu";

import classes from "./LabelItem.module.css";
const LabelItem = (props) => {
  const inputRef = useRef();
  const dispatch = useDispatch();

  const [options, showOptions] = useState(false);
  const [disabled, setDisabled] = useState(true);
  const [saveBtn, setSaveBtn] = useState(false);

  // const labelHandler = useMemo(() => {
  //   console.log("inside the label function");
  //   dispatch(todoActions.groupByLabel(props.keyOfLabel));
  // }, [props.keyOfLabel]);

  const labelHandler = () => {
    dispatch(todoActions.groupByLabel(props.keyOfLabel));
  };

  const removeLabelHandler = (id) => {
    showOptions(!options);
    dispatch(
      deleteLabel({
        key: props.keyOfLabel,
      })
    );

    dispatch(todoActions.setBacktoAllItems());
  };

  const showOptionsHandler = () => {
    showOptions(!options);
  };

  const saveLabelnameHandler = () => {
    setSaveBtn(false);
    setDisabled(true);
    dispatch(
      updateLabel({
        name: inputRef.current.value,
        key: props.keyOfLabel,
        id: props.id,
      })
    );
  };
  const updateLabelHandler = (label) => {
    showOptions(!options);
    setDisabled(false);
    setSaveBtn(true);
  };

  return (
    <React.Fragment>
      <div className={classes.labelParent}>
        <div className={classes.fullLabel}>
          <div className={classes.label}>
            <i
              className={`fa-solid fa-ellipsis-vertical ${classes.options}`}
              onClick={(e) => showOptionsHandler(props.title)}
            ></i>
            <div onClick={labelHandler}>
              <input
                ref={inputRef}
                className={classes.labelButton}
                defaultValue={props.title}
                disabled={disabled}
              />
            </div>
          </div>
          <div className={classes["arrow-right"]} />
          {saveBtn && (
            <div className={classes.saveBtn}>
              <Button
                size="small"
                variant="outlined"
                onClick={saveLabelnameHandler}
              >
                Save
              </Button>
            </div>
          )}
        </div>
      </div>

      {options && (
        <IconMenu
          keyOfLabel={props.keyOfLabel}
          removeLabelHandler={removeLabelHandler}
          updateLabelHandler={updateLabelHandler}
        />
      )}
    </React.Fragment>
  );
};

export default LabelItem;
