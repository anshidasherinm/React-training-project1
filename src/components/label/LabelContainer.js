import React, { useEffect } from "react";
import LabelItem from "components/label/LabelItem";
import { getLabelList } from "store/label/label-thunk";
import { todoActions } from "store/todo/todo-slice";
import { useAppDispatch, useAppSelector } from "store/redux-hooks";
import { Button, Card, List } from "@mui/material";
import { makeStyles } from "@material-ui/core/styles";

const useStyles = makeStyles({
  container: {
    height: "240px",
    width: "310px",
    backgroundColor: "white",
    overflowY: "auto",
    marginLeft: "1rem",
    paddingBottom: "1rem",
    position: "relative",
  },

  allBtn: {
    marginTop: "2rem",
    marginBottom: "0rem",
    marginLeft: "1rem",
    width: "100px",
  },

  emptyList: {
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    width: "100%",
    height: "8rem",
    color: "white",
  },
});
const LabelContainer = (props) => {
  const labelItems = useAppSelector((state) => state.label.labels);
  console.log(labelItems);
  const dispatch = useAppDispatch();
  const classes = useStyles();
  useEffect(() => {
    dispatch(getLabelList());
  }, []);

  const showAllHandler = () => {
    dispatch(todoActions.setBacktoAllItems());
  };

  return (
    <Card
      style={{
        width: "90%",
        backgroundColor: "#383838",
      }}
    >
      <div className={classes.allBtn}>
        <Button
          size="large"
          onClick={showAllHandler}
          variant="outlined"
          sx={{
            backgroundColor: "orange",
            borderRadius: "3rem",
            border: "none",
            color: "white",
            fontWeight: "bold",
            ":hover": {
              backgroundColor: "#ffb347",
              color: "white",
              border: "none",
            },
          }}
        >
          All
        </Button>
      </div>

      {labelItems.length > 0 && (
        <List
          sx={{
            width: "100%",
            paddingTop: "2rem",
            paddingBelow: "2rem",
            marginTop: "1rem",
            height: "10rem",
            overflowY: "auto",
            display: "flex",
            flexDirection: "column",
            alignItems: "center",
          }}
        >
          {labelItems.map((label) => {
            return (
              <LabelItem
                data-testid={"input-labelitem"}
                id={label.id}
                title={label.name}
                key={label.key}
                keyOfLabel={label.key}
              />
            );
          })}
        </List>
      )}
      {labelItems.length < 1 && (
        <div className={classes.emptyList}>No Labels </div>
      )}
    </Card>
  );
};

export default LabelContainer;
