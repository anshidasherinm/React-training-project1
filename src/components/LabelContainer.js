import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { getLabelList } from "../store/label/label-thunk";
import { todoActions } from "../store/todo/todo-slice";
import classes from "./LabelContainer.module.css";
import LabelItem from "./LabelItem";
import { Button } from "@mui/material";

const LabelContainer = () => {
  const labelItems = useSelector((state) => state.label.labels);
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(getLabelList());
  }, []);
  const showAllHandler = () => {
    dispatch(todoActions.setBacktoAllItems());
  };
  return (
    <div className={classes.container}>
      <div className={classes.allBtn}>
        <Button size="large" onClick={showAllHandler} variant="outlined">
          All
        </Button>
      </div>
      {labelItems.map((label) => {
        return (
          <LabelItem
            id={label.id}
            title={label.name}
            key={label.key}
            keyOfLabel={label.key}
          />
        );
      })}
    </div>
  );
};

export default React.memo(LabelContainer);
