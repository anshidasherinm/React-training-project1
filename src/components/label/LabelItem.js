import React, { useMemo, useRef, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { deleteLabel, updateLabel } from "../../store/label/label-thunk";
import { todoActions } from "../../store/todo/todo-slice";
import {
  Badge,
  Button,
  Card,
  Checkbox,
  Chip,
  Grid,
  IconButton,
  List,
  ListItem,
  ListItemButton,
  ListItemText,
  TextField,
} from "@mui/material";
import DeleteIcon from "@mui/icons-material/Delete";
import EditIcon from "@mui/icons-material/Edit";

import ListItemIcon from "@mui/material/ListItemIcon";

import classes from "./LabelItem.module.css";
const LabelItem = (props) => {
  const inputRef = useRef();
  const dispatch = useDispatch();

  const [options, showOptions] = useState(false);
  const [disabled, setDisabled] = useState(true);
  const [saveBtn, setSaveBtn] = useState(false);

  const todoItems = useSelector((state) =>
    state.todo.allItems.filter((todo) => todo.labelKey === props.keyOfLabel)
  );
  const count = todoItems.length;
  const labelHandler = () => {
    dispatch(todoActions.groupByLabel(props.keyOfLabel));
  };

  const removeLabelHandler = (id) => {
    showOptions(!options);
    dispatch(
      deleteLabel({
        key: props.keyOfLabel,
      })
    );

    dispatch(todoActions.setBacktoAllItems());
  };

  const saveLabelnameHandler = () => {
    setSaveBtn(false);
    setDisabled(true);
    dispatch(
      updateLabel({
        name: inputRef.current.value,
        key: props.keyOfLabel,
        id: props.id,
      })
    );
  };

  const editLabelHandler = () => {
    setSaveBtn(true);
    setDisabled(false);
  };

  return (
    <ListItem
      style={{
        height: "2rem",
        marginBottom: "1rem",
        color: "white",
        display: "flex",
        alignItems: "center",
        width: "100%",
        justifyContent: "center",
      }}
      key={props.keyOfLabel}
      secondaryAction={
        <IconButton edge="end" aria-label="comments"></IconButton>
      }
      disablePadding
    >
      <Grid
        container
        spacing={2}
        style={{
          // height: "3rem",
          padding: "0",
          display: "flex",
          alignItems: "center",
          justifyContent: "center",
          margin: "0 1rem",
          width: "100%",
        }}
      >
        <Grid
          item
          xs={6}
          md={8}
          style={{
            backgroundColor: "white",
            // backgroundColor: "red",
            borderRadius: "5rem",
            display: "flex",
            justifyContent: "center",
            alignItems: "center",
            padding: "0",
            height: "2rem",
          }}
        >
          <Badge
            badgeContent={count}
            color="success"
            sx={{
              "& .MuiBadge-badge": {
                color: "white",
                backgroundColor: "orange",
              },
              width: "70%",
            }}
          >
            <ListItemButton
              role={undefined}
              onClick={disabled && labelHandler}
              dense
              sx={{
                height: "2rem",
              }}
            >
              <p id="edit" className={classes.titleOfLabel}>
                <input
                  ref={inputRef}
                  className={classes.labelButton}
                  defaultValue={props.title}
                  disabled={disabled}
                />
              </p>
            </ListItemButton>
          </Badge>
        </Grid>
        <Grid
          item
          xs={3}
          md={4}
          style={{
            display: "flex",
            justifyContent: "space-evenly",
            alignItems: "center",
            padding: "0",
            height: "2rem",
            // backgroundColor: "orange",
          }}
        >
          {saveBtn === false && (
            <React.Fragment>
              <EditIcon
                sx={{
                  marginLeft: "2rem",
                  padding: "0.2rem",
                  color: "green",
                  borderRadius: "5rem",
                  backgroundColor: "rgb(161, 236, 161)",
                }}
                onClick={editLabelHandler}
              />
              <DeleteIcon
                onClick={removeLabelHandler}
                sx={{
                  color: "rgba(245, 31, 31, 0.911)",
                  backgroundColor: "rgb(231, 108, 92)",
                  borderRadius: "5rem",
                  marginLeft: "0.5rem",
                  marginRight: "1rem",
                  padding: "0.2rem",
                }}
              />
            </React.Fragment>
          )}
          {saveBtn === true && (
            <ListItemIcon>
              <Button
                size="small"
                onClick={saveLabelnameHandler}
                sx={{
                  marginLeft: "1rem",
                  backgroundColor: "orange",
                  borderRadius: "5rem",
                  color: "white",
                  fontWeight: "large",
                  ":hover": { backgroundColor: "#ffb347", color: "white" },
                }}
              >
                Save
              </Button>
            </ListItemIcon>
          )}
        </Grid>
      </Grid>
    </ListItem>
  );
};

export default LabelItem;
