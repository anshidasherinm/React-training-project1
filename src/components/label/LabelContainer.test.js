import { render, screen } from "@testing-library/react";
import { useAppDispatch, useAppSelector } from "store/redux-hooks";
import { testUseAppSelector } from "store/test-app-selector";
import "@testing-library/jest-dom";
import * as redux from "react-redux";
import { shallow } from "enzyme";
import LabelContainer from "components/label/LabelContainer";
import LabelItem from "components/label/LabelItem";

jest.mock("store/redux-hooks");
describe("LabelContainer", () => {
  beforeEach(() => {
    useAppSelector.mockImplementation(testUseAppSelector);
    // const dispatch = jest.fn();
    useAppDispatch.mockReturnValue(jest.fn());
  });
  //   it("renders No Labels", () => {
  //     //Arrange
  //     render(<LabelContainer />);

  //     //Act
  //     //...nothing

  //     //Assert
  //     const emptyList = screen.getByText("No Labels");
  //     expect(emptyList).toBeInTheDocument();
  //   });

  // it("renders No Labels", () => {
  //   //Arrange
  //   render(<LabelContainer />);
  //   const useSelectorMock = jest
  //     .spyOn(redux, "useSelector")
  //     .mockReturnValueOnce([
  //       // { id: 1463, key: "-NOqXGK_kreZB5AW_PFG", name: "School" },
  //     ]);

  //   //Act
  //   //...nothing

  //   //Assert
  //   const emptyList = screen.getByText("No Labels");
  //   expect(emptyList).toBeInTheDocument();
  // });

  it("renders a label", async () => {
    render(<LabelContainer />);
    const enzymeWrapper = shallow(
      <LabelItem id={1463} key="-NOqXGK_kreZB5AW_PFG" name="School" />
    );
    expect(enzymeWrapper.text()).toMatchSnapshot();
  });
});
