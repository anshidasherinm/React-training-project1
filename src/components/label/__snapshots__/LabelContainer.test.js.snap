// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelContainer renders a label 1`] = `""`;
